### PR TITLE
logging RFC draft

### PR DESCRIPTION
**Pull request by TravisEz13**

### CODE REVIEW

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+    As a administrator of machines which run PowerShell (machine admins),
+    I can easily log stript blocks which PowerShell executes to systems such as Azure Log Analytics or Splunk (log systems),
+    so that I can analyze PowerShell activity on all the machines I run.
+
+## User Experience
+
+This example assumes you have registered the key to communicate with Azure Log Analytics with PowerShell Secrets with the name `myKeyName`.
+
+```powershell
```
Comment (SydneyhSmith) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md:55]: Link to the secret management RFC

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+Register-PSScriptBlockLogger -Provdier AzureLogAnalytics -KeyName myKeyName -Properties @{
+    WorkspaceId = ‘myworkspaceid’
+}
+```
+
+```output
```
Comment (SydneyhSmith) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md:61]: -Get and -Unregister cmdlets could be helpful

What is the error message or warning if you try to register a second, a -Force parameter?

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+1. The transport to the Logging system is secure.
+   Secure includes:
+   * Encrypted
+   * Authenticated
+1. The machine admin is willing and authorized to send all ScriptBlocks to the logging system.
+
+### Registration
+
+Only one logging provider can be registered at a time.
+There will be two options for registration.
```
Comment (SydneyhSmith) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: The two options for registration are explained below

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+and delete items if the number of items exceeds a predefined limit.
+
+### Data to send
+
+| Name                  | Type     | Description                                                                            |
+|-----------------------|----------|----------------------------------------------------------------------------------------|
+| ScriptBlockText       | string   | The text of the script block (may be truncated, based on field size limit of provider) |
+| UtcTime               | datetime | The time on the local machine which the ScriptBlock was started                        |
+| ScriptBlockHash       | string   | A hash of the script block                                                             |
+| File_s                | string   | If applicable, the name of the file the script block came from                         |
```
Comment (SydneyhSmith) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: Remove the _s

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+to walk you through creating an Azure Log Analytics workspaces and creating a deployment script.
+
+### API used to send data
+
+https://docs.microsoft.com/en-us/azure/azure-monitor/platform/data-collector-api
+This API is in preview and is subject to change.
+This is one of the reasons for the feature being experimental.
+
+## Alternate Proposals and Considerations
+
```
Comment (SydneyhSmith) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md:225]: Should probably link to or mention the original RFC

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+SupersededBy:
+Version: 1.0
+Area: Engine
+Comments Due: 08/15/2019
+Plan to implement: Yes
+---
+
+# Title
+
+ScriptBlock Logging
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
Remote ScriptBlock Logging
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+## Description
+
+This is an addition to the existing ScriptBlock creation logging.
+The goal is to make it easier to configure to collect the logs centrally.
+This feature will be experimental.
+
+## Motivation
+
+As a administrator of machines which run PowerShell (machine admins),
+I can easily log stript blocks which PowerShell executes to systems such as Azure Log Analytics or Splunk (log systems),
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
I can easily log script blocks which PowerShell executes to systems such as Azure Log Analytics or Splunk (log systems),
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+---
+
+# Title
+
+ScriptBlock Logging
+
+## Description
+
+This is an addition to the existing ScriptBlock creation logging.
+The goal is to make it easier to configure to collect the logs centrally.
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md:19]: I think it's worth calling out why remote logging is desirable.  If the machine gets pwned, an attacker can just clear the local logs.

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+```output
+Error: A provider is already registered.  User -Force to overwrite the current provider settings.
+```
+
+### Unregister
+
+```powershell
+Unregister-PSScriptBlockLogger
+}
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: This example is incomplete

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+```
+
+There would be no output.
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
Upon success, there would be no output.
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+### Registration
+
+Only one logging provider can be registered at a time.
+There will be two options for registration,
+which are explained below:
+
+#### Registration cmdlet
+
+There will be a cmdlet to register a logging provider.
+`Register-PSScriptBlockLogger` will allow you to register a provider.
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: `Provider` can be confusing as I think most people think of PowerShell Provider.  Perhaps use the word extension.  I don't see where you define the model for extensions.  Perhaps also show the user experience if 3rd parties publish extensions to PSGallery and how they get registered.

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md (hunk 9 above)
Comment (TravisEz13) [reply 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+# Title
+
+Remote ScriptBlock Logging
+
+## Description
+
+This is an addition to the existing ScriptBlock creation logging.
+The goal is to make it easier to configure to collect the logs centrally.
+Because many ScriptBlocks contain sensitive information,
+it is advised to encrypt the logs locally against attacker that might compromise the local machine.
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md:21]: We're not doing the encrypted logging for this iteration, I think it's better to call out the remote aspects of the logging as a security measure

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 13]
```diff
@@ -0,0 +1,280 @@
...
+        ScriptBlockLogging  = @{
+            TypeName='TypeName'
+        }
+    }
+}
+```
+
+The class must have a constructor with a parameter of type SecureString to pass the key and parameters for the rest of the Properties in `ScriptBlockLoggingExtensionProperties`.
+
+The class must have a void method named `PostScriptBlockLog` with the following parameters:
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: I've always thought of `log` as the entire thing and `trace` as an entry in the log, so `PostScriptBlockTrace` seems better or perhaps `PostScriptBlogLogEntry`

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md [hunk 14]
```diff
@@ -0,0 +1,280 @@
...
+   - The first command name in the ScriptBlock
+1. `runspaceId` of type int
+   - a numeric identifier of the runspace the ScriptBlock was created in.
+1. `runspaceName` of type string
+   - The name the of the runspace the ScriptBlock was created in, if set.
+
+We will add code in `CompiledScriptBlock.LogScriptBlockCreation`,
+which does the work to load and call the extension.
+
+PowerShell will not retry calls to the extension which fail.
```
Comment (SteveL-MSFT) [inline 1-Draft/RFCXXXX-ScriptBlockLogging.md]: We should log locally if an extension fails.  Do we no longer retry forever?

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md (hunk 14 above)
Comment (TravisEz13) [reply 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
PowerShell will not retry calls to the extension which fail and will log locally that the call failed.
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md (hunk 13 above)
Comment (TravisEz13) [reply 1-Draft/RFCXXXX-ScriptBlockLogging.md]:
```suggestion
The class must have a void method named `PostScriptBlockEntry` with the following parameters:
```

File: 1-Draft/RFCXXXX-ScriptBlockLogging.md (hunk 12 above)
Comment (TravisEz13) [reply 1-Draft/RFCXXXX-ScriptBlockLogging.md:21]: It’s two lines down.

Comment (TravisEz13) [reply 1-Draft/RFCXXXX-ScriptBlockLogging.md:21]: This is part of calling this out.  Pointing out the disadvantages of logging locally

**Pull request closed without merging**